dbms/src/Functions/FunctionsJson.h

```diff
@@ -530,11 +612,671 @@ class FunctionsJsonArray : public IFunction
+    bool useDefaultImplementationForConstants() const override { return true; }
+
+    void setInputTiDBFieldType(const tipb::FieldType & tidb_tp_) { input_tidb_tp = &tidb_tp_; }
+    void setOutputTiDBFieldType(const tipb::FieldType & tidb_tp_) { output_tidb_tp = &tidb_tp_; }
+    void setCollator(const TiDB::TiDBCollatorPtr & collator_) override { collator = collator_; }
+
+    DataTypePtr getReturnTypeImpl(const DataTypes & arguments) const override
+    {
+        const auto & input_type = arguments[0];
+        if (input_type->onlyNull())
+        {
+            return input_type;
+        }
+
+        if unlikely (!removeNullable(input_type)->isStringOrFixedString())
+            throw Exception(
+                fmt::format("Illegal type {} of argument of function {}", arguments[0]->getName(), getName()),
+                ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT);
+        auto return_type = std::make_shared<DataTypeString>();
+        return input_type->isNullable() ? makeNullable(return_type) : return_type;
+    }
+
+    void executeImpl(Block & block, const ColumnNumbers & arguments, size_t result) const override
+    {
+        const auto & from = block.getByPosition(arguments[0]);
+        if (from.type->onlyNull())
+        {
+            auto only_null_column = from.column;
+            block.getByPosition(result).column = std::move(only_null_column);
+            return;
+        }
+
+        auto nested_block = createBlockWithNestedColumns(block, arguments);
+        auto input_source = createDynamicStringSource(*nested_block.getByPosition(arguments[0]).column);
+
+        auto col_to = ColumnString::create();
+        auto & data_to = col_to->getChars();
+        auto & offsets_to = col_to->getOffsets();
+        auto rows = block.rows();
+        offsets_to.resize(rows);
+
+        // In raw function test, input_tidb_tp/output_tidb_tp is nullptr.
+        if (collator && collator->isBinary())
+        {
+            if (unlikely(input_tidb_tp == nullptr))
+            {
+                doExecuteForBinary<false>(data_to, offsets_to, input_source, input_tidb_tp->tp(), -1, block.rows());
+            }
+            else if (input_tidb_tp->tp() == TiDB::TypeString)
+            {
+                doExecuteForBinary<true>(
+                    data_to,
+                    offsets_to,
+                    input_source,
+                    input_tidb_tp->tp(),
+                    input_tidb_tp->flen(),
+                    block.rows());
+            }
+            else
+            {
+                doExecuteForBinary<false>(
+                    data_to,
+                    offsets_to,
+                    input_source,
+                    input_tidb_tp->tp(),
+                    input_tidb_tp->flen(),
+                    block.rows());
+            }
+        }
+        else if ((unlikely(output_tidb_tp == nullptr)) || hasParseToJSONFlag(*output_tidb_tp))
```

Comment:
Do we have integration tests to test that TiDB will set this ParseToJSON flag correctly?

Reply:
yes, https://github.com/pingcap/tiflash/pull/8333/files#diff-808e483af0ccfb82cb5b86947a1f11b3dac0b6704897bffb4e3cc90ff53ec23aR253-R266